### PR TITLE
add hook support. Closes #68

### LINKED ISSUE #68: Hook support
Might be nice to have some hooks for builds, linting, running tests, whatever. We could potentially move some things like the shimmed language build commands to this instead, as a default, and remove part of the `runtime.Runtime` interface.

We could then remove `runtime.Runtime` all together if we wanted, and just provide defaults for these properties. Shimmed() would be an option, Handler() an option, DefaultFile() an option, Build() would be a hook, Clean() would be a hook.

This would allow people to extend Apex without contributing Go code in core as well. For example you might even want to shim node code to act over stdio instead, just because it's nicer to work with.

We could also move the Go runtime portion to a different package then, just so they're not conflated, just apex/runtime. 

- [ ] remove runtime.Runtime
- [ ] define defaults
- [ ] refactor how inference works (define the default file detection stuff elsewhere)
- [ ] add "shim" (bool)
- [ ] add "handler" (string)
- [ ] add "hooks" (map of strings)
- [ ] add hooks for pre/post-deploy and others

Pretty sold on this I think. We shouldn't really focus on runtimes, they'll add Go support eventually and same with the others. These could also be inherited from the project level, which would be useful in most cases, for linting, pre-compiling with Babel etc.

### PR DESCRIPTION
**Pull request by tj**
Only adds two hooks for now:

- build (pre-zip)
- clean (post-zip)

We can add more on request, pre/post-deploy might be good for linting/testing but I'm using build and clean here so that the `apex build` command works as well

- [ ] update wiki

### CODE REVIEW

File: function/function.go [hunk 1]
```diff
@@ -449,20 +466,64 @@ func (f *Function) Zip() (io.Reader, error) {
...
+// Error string.
+func (e *HookError) Error() string {
+	return fmt.Sprintf("hook %q: %s", e.Name, e.Output)
+}
+
+// RunHook executes hook `name` in the function's directory.
+func (f *Function) RunHook(name string) error {
+	var command string
+
+	switch name {
```
Comment (tj) [inline function/function.go]: this is kinda lame but executing them unnamed in the log output could be confusing

Comment (mthenw) [inline function/function.go]: fine for now imho

**Comment by mthenw** [comment]
This is more than great. Thanks!

**Comment by tj** [comment]
woot thanks for the review!

**Pull request merged**